### DISCUSSION TITLE
Assistance with extending classes with Loggable and ClassMethods

### DISCUSSION THREAD

**Discussion by todd-a-jacobs** [reidmorrison/semantic_logger#209] [Q&A]
I'm trying to use SemanticLogger 4.10.0 as a part of a dynamic tracing solution in a Ruby 3.1.1 app. However, I seem to be misunderstanding something about how or where to access *logger* and #logger_measure_method because I can't seem to implement them through extending the class.

```ruby
require 'json'
require 'semantic_logger'

module SomeGem
  module CustomLogger
    include SemanticLogger

    class MyFormat < SemanticLogger::Formatters::Default
      def call (log, logger, machine_name: nil, json: nil)
        self.log = log
        self.logger = logger
        message = "[#{machine name}] #{message}" if machine_name
        payload = JSON.parse(json)
        [time, level, process.info, tags, named.tags, duration, name, message, payload, exception].compact!
      end
    end

    def self.extended(klass)
      prepend SemanticLogger, SemanticLogger::Loggable, SemanticLogger::Loggable::ClassMethods

      SemanticLogger.default_level = :debug
      SemanticLogger.add_appender(file_name: "/dev/stderr", formatter: SomeGem::CustomLogger::MyFormat.new,
                                  level: :warn, filter: proc { %i[warn fatal unknown].include? _1.level }
      )
      SemanticLogger.add_appender(file_name: "/dev/stdout", formatter: SomeGem::CustomLogger::MyFormat.new,
                                  level: :trace, filter: proc { %i[trace debug info error].include? _1.level }
      )

      define_method(:auto_measure_method_calls) do
        public_instance_methods(false).
          reject { _1.to_s.match? /^logger/ }.
          each { logger_measure_method _1, level: :trace }
      end
   end
  end
end

module SomeGem
  class Foo
    extend CustomLogger

    def alpha   = 1
    def beta    = 2
    def charlie = 3

    auto_measure_method_calls if ENV['LOG_LEVEL'] == 'trace'
  end
end

ENV['LOG_LEVEL'] = 'trace'
f = SomeGem::Foo.new
pp f.alpha, f.beta, f.charlie
```

My issues are:

1. The underlying X/Y issue is that I want to DRY up the code of including SemanticLogger::Loggable and  SemanticLogger::Loggable::ClassMethods in every class I'm tracing, and auto-measuring its methods in development.
2. When I try to extend a class with a module that is intended to pull in SemanticLogger::Loggable, I don't seem to always have *logger* available as an accessor throughout the class.
3. I'm also concerned that including the module in multiple classes would result in duplicate appenders being added to the #appenders array, wherever that's actually stored.
4. Most importantly though, when I try to automagically add logging and method measurement through extended a class, I get errors like the following:

    > ~/.gem/ruby/3.1.1/gems/semantic_logger-4.10.0/lib/semantic_logger/loggable.rb:96:in \`alpha': undefined local variable or method \`logger' for #<SomeGem::Foo:0x00000001138f6a38> (NameError) from foo.rb:51:in \`\<main>'

Am I missing something obvious? If there's a better way to accomplish what I'm trying to do, that's great! If I'm doing something wrong, understanding that would be great, too. If it's a bug, or I'm using the feature wrong, that's useful information as well.

P.S. I've posted [a slightly-expanded version of this question](https://stackoverflow.com/q/71298654/1301972) on StackOverflow. It's mostly "expanded" to provide additional context for those unfamiliar with SemanticLogger and to break the question up into sections to fit limitations of SO's Markdown rendering. Still, if anyone wants SO rep for answering there instead, I wanted to provide a link.

**Comment by reidmorrison** [comment]
This is an example of how I have included `SemanticLogger::Loggable` into a concern / mixin: https://github.com/reidmorrison/rocketjob/blob/master/lib/rocket_job/subscriber.rb

Semantic Logger follows a design pattern created by log4j where every class has its own logger instance. It includes things like the class name. The appenders are completely separate and all loggers forward their messages to the appenders.

For a design overview of Semantic Logger, scroll down to the images on https://logger.rocketjob.io/custom_appenders.html

**Reply by todd-a-jacobs** [reply]
I'll carve out some time today to look at that. I may generate some flow or class diagrams, too, as that may help with understanding (for me and for others).

**Comment by todd-a-jacobs** [comment]
I need to write a more comprehensive answer to my own question which involves calling #class_exec when the module is included, but as a "quick and dirty" way to introspect, it turns out that SemanticLogger creates a *@semantic_logger* instance variable for each class where SemanticLogger::Loggable is included. I figure that information may help someone until I can post a more complete answer.
